source/s3_client.c

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "s3_client.h"
+
+#include "auth.h"
+#include "io.h"
+
+static const char *s_capsule_name_s3_client = "aws_s3_client";
+
+struct s3_client_binding {
+    struct aws_s3_client *native;
+
+    bool release_called;
+    bool shutdown_called;
+
+    /* Shutdown callback, reference cleared after setting result */
+    PyObject *on_shutdown;
+};
+
+static void s_s3_client_release(struct s3_client_binding *client) {
```

Comment:
trivial: this fn is only ever called from the capsule_destructor. why not just combine the functions?